pkg/k8s/portforward.go

```diff
@@ -0,0 +1,149 @@
+package k8s
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+
+	// Load all the auth plugins for the cloud providers.
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+)
+
+// PortForward provides a port-forward connection into a Kubernetes cluster.
+type PortForward struct {
+	method     string
+	url        *url.URL
+	localPort  int
+	remotePort int
+	stopCh     chan struct{}
+	readyCh    chan struct{}
+	config     *rest.Config
+}
+
+// NewPortForward returns an instance of the PortForward struct that can be used
+// to establish a port-forward connection to a pod in the deployment that's
+// specified by namespace and deployName. If localPort is 0, it will use a
+// random ephemeral port.
+func NewPortForward(
+	configPath, kubeContext, namespace, deployName string,
+	localPort, remotePort int,
+) (*PortForward, error) {
+	config, err := GetConfig(configPath, kubeContext)
+	if err != nil {
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	kubeAPI := &KubernetesAPI{Config: config}
+	client, err := kubeAPI.NewClient()
+	if err != nil {
+		return nil, err
+	}
+
+	pods, err := kubeAPI.GetPodsByNamespace(client, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	podName := ""
+	for _, pod := range pods {
+		if pod.Status.Phase == v1.PodRunning {
+			if strings.HasPrefix(pod.Name, deployName) {
+				podName = pod.Name
+				break
+			}
+		}
+	}
+
+	if podName == "" {
+		return nil, fmt.Errorf("no running pods found for %s", deployName)
+	}
+
+	req := clientset.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Namespace(namespace).
+		Name(podName).
+		SubResource("portforward")
+
+	if localPort == 0 {
+		localPort, err = getLocalPort()
```

Comment:
is there a potential timing issue where we call `getLocalPort` in `NewPortForward`, and then that `localPort` is taken before `Run()` is called?

is the reason we're calling `getLocalPort` here to enable `URLFor` to work immediately after calling `NewPortForward`?